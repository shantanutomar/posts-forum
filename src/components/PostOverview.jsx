import React from 'react';
import Paper from '@material-ui/core/Paper';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles((theme) => ({
  paperRoot: {
    display: 'flex',
    flexWrap: 'wrap',
    width: '50%',
    '& > *': {
      margin: theme.spacing(1),
      width: '100%',
      padding: '12px',
      backgroundColor: '#006a71',
      color: '#ffffdd'
    },
    '@media (max-width: 450px)': {
      width: '100%',
      margin: 0
    }
  },
  viewOnlyPaperRoot: {
    width: '100%',
  },
  postTitle: {
    fontSize: '20px',
    fontWeight: '500',
  },
  userName: {
    cursor: 'pointer',
    marginTop: '15px',
  }
}));

const PostOverview = (props) => {
  const classes = useStyles();

  return (
    <div key={props.post.id} className={!props.viewOnly ? classes.paperRoot : 
        `${classes.paperRoot} ${classes.viewOnlyPaperRoot}`} 
      onClick={!props.viewOnly ? (event) => props.onBlockClick(event, 'post', props.post.id) : null}>
      <Paper elevation={3}>
        <div className={classes.postTitle}>{props.post.title}</div> 
        <div className={classes.userName} 
          onClick={!props.viewOnly ? (event) => props.onBlockClick(event, 'user', props.post.userId) : null}>
          Created By {props.users[props.post.userId].username}</div>
      </Paper>
    </div>
  );
}

export default PostOverview;