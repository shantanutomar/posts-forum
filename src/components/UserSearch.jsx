import React from 'react';
import TextField from '@material-ui/core/TextField';
import Autocomplete from '@material-ui/lab/Autocomplete';
import { connect } from 'react-redux';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles((theme) => ({
  userSearchCont: {
    width: '90%',
    margin: '10px auto',
    zIndex: '1',
  },
  autoCompleteRoot: {
    width: '100%',
    margin: 'auto',
    backgroundColor: '#fff',
  }
}));

const UserSearch = props => {
  const classes = useStyles();

  return (
    <section className={classes.userSearchCont}>
      {!props.users.isUsersFetching ? 
        <Autocomplete
        value={props.value || null}
        options={Object.values(props.users.users).map(user => user.username)}
        className={classes.autoCompleteRoot}
        onChange={(event, newValue) => props.onSearchInputChange(newValue)}
        renderInput={(params) => <TextField {...params} label="Search Users" variant="outlined" />}
      /> : null}
    </section>
  );
}
const mapStateToProps = state => {
  return {
    users: state.usersReducer
  }
}

export default connect(mapStateToProps, null)(UserSearch);