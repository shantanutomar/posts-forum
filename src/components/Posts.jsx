import React, { useEffect, useRef } from "react";
import { makeStyles } from '@material-ui/core/styles';
import { connect } from 'react-redux';
import { fetchPostsRequest, fetchInitialPostsAndUsers, fetchPosts } from "../actions/postsActions";
import { fetchUsersRequest } from "../actions/userActions";
import PostsAndUserViews from "./PostsAndUserViews";
import UserSearch from "./UserSearch";
import PostOverview from "./PostOverview"
import CircularProgress from '@material-ui/core/CircularProgress';

const useStyles = makeStyles(theme => ({
  postsCont: {
    padding: '0 5%',
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'center',
    '@media (max-width: 450px)': {
      padding: '0 2%',
    }
  },
  loader: {
    textAlign: 'center',
    marginTop: '60px',
  }
}));

const Posts = props => {
  const classes = useStyles(props);
  const posts = useRef();
  const users = useRef();
  const searchValueRef = useRef();
  const [openDialog, setOpenDialog] = React.useState(false);
  const [dialogProps, setDialogProps] = React.useState({ selectedView: 'post', id: 0 });
  const [searchValue, setSearchValue] = React.useState(null);

  useEffect(() => {
    const handleScroll = () => {

      if((window.innerHeight + document.documentElement.scrollTop) <= document.documentElement.offsetHeight - 1) return;      

      if(Object.values(posts.current).length < 100 && !searchValueRef.current) {
        setTimeout(() => {
          props.fetchPostsRequest();
          props.fetchPosts(Object.values(posts.current).length);  
        }, 300)
      }
    }
    
    props.fetchPostsRequest();
    props.fetchUsersRequest();
    props.fetchInitialPostsAndUsers();
    window.addEventListener('scroll', handleScroll);
    return () => window.removeEventListener('scroll', handleScroll);
  }, []);
  
  useEffect(() => {
    posts.current = props.posts.posts;
    users.current = props.users.users;
    searchValueRef.current = searchValue;
  });

  const handleClose = () => {
    setOpenDialog(false);
  };

  const onSearchInputChange = searchValue => {
    searchValue ? setSearchValue(searchValue) : setSearchValue(null); 
  }

  const onBlockClick = (event, selectedView, id) => {
    event.stopPropagation();
    setOpenDialog(true);
    setDialogProps({ selectedView, id });
  }

  const filterPostsResults = () => {
    let renderPosts = null;
    if(searchValue) {
      const selectedUser = Object.values(props.users.users).filter(user => user.username === searchValue);
      renderPosts = Object.values(props.posts.posts).filter(post => post.userId === selectedUser[0].id).map(post => {
        return <PostOverview key={post.id} onBlockClick={onBlockClick} post={post} users={props.users.users} />        
      });
    } else {
      renderPosts = Object.values(props.posts.posts).map(post => {
        return <PostOverview key={post.id} onBlockClick={onBlockClick} post={post} users={props.users.users} />
      });  
    }
    return renderPosts.length > 0 ? renderPosts : 'Posts Not Available';
  }

  return (
    props.users.isUsersFetching && !props.posts.isPostsFailure ? 
    <div className={classes.loader}><CircularProgress /></div> :
    <section className={classes.main}>
      <UserSearch onSearchInputChange={onSearchInputChange} value={searchValue} />
      <section className={classes.postsCont}>
        {!props.posts.isPostsFailure ? filterPostsResults() : 'Something went wrong'}
      </section>
      {props.posts.isPostsFetching ? <div>Loading More Posts...</div> : null}
      <PostsAndUserViews open={openDialog} handleClose={handleClose} { ...dialogProps } />
    </section>
  );
}

const mapStateToProps = state => {
  return {
    posts: state.postsReducer,
    users: state.usersReducer
  }
}

const mapDispatchToProps = dispatch => {
  return {
    fetchPostsRequest: () => dispatch(fetchPostsRequest()),
    fetchPosts: (start) => dispatch(fetchPosts(start)),
    fetchUsersRequest: () => dispatch(fetchUsersRequest()),
    fetchInitialPostsAndUsers: () => dispatch(fetchInitialPostsAndUsers()),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Posts);