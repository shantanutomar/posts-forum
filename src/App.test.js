import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import PostOverview from "./components/PostOverview";
import Posts from "./components/Posts";
import UserSearch from "./components/UserSearch";

const mockStore = configureStore([]);

describe('React-Redux Component', () => {
  let store;
  let PostOverviewRef; 
  let PostsRef;
  let UserSearchRef;
 
  beforeEach(() => {
    store = mockStore({
      postsReducer: {
        posts: {1: {body: 'Test posts body', id: 1, title: 'Test posts title', userId: 1}},
        isPostsFetching: false,
        isPostsSuccess: false,
        isPostsFailure: false,
        isPostCommentsFetching: false,
        isPostCommentsSuccess: false,
        isPostCommentsFailure: false,  
      },
      usersReducer: {
        users: {1: {email: "Sincere@april.biz", id: 1, name: "Leanne Graham", phone: "1-770-736-8031 x56442", 
        username: "Bret", website: "hildegard.org", company: {name: "Romaguera-Crona"}}},
        isUsersFetching: false,
        isUsersSuccess: false,
        isUsersFailure: false    
      }
    });

    PostOverviewRef = renderer.create(
      <Provider store={store}>
        <PostOverview post={store.getState().postsReducer.posts[1]} users={store.getState().usersReducer.users}/>
      </Provider>
    );
    PostsRef = renderer.create(
      <Provider store={store}>
        <Posts />
      </Provider>
    );
    UserSearchRef = renderer.create(
      <Provider store={store}>
        <UserSearch />
      </Provider>
    );
  });
 
  it('PostOverview should render with given state from Redux store', () => {
    expect(PostOverviewRef.toJSON()).toMatchSnapshot();
  });

  it('Posts should render with given state from Redux store', () => {
    expect(PostsRef.toJSON()).toMatchSnapshot();
  });

  it('UserSearch should render with given state from Redux store', () => {
    expect(UserSearchRef.toJSON()).toMatchSnapshot();
  });
});