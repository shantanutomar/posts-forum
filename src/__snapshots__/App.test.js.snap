// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component PostOverview should render with given state from Redux store 1`] = `
<div
  className="makeStyles-paperRoot-1"
  onClick={[Function]}
>
  <div
    className="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
  >
    <div
      className="makeStyles-postTitle-3"
    >
      Test posts title
    </div>
    <div
      className="makeStyles-userName-4"
      onClick={[Function]}
    >
      Created By 
      Bret
    </div>
  </div>
</div>
`;

exports[`My Connected React-Redux Component Posts should render with given state from Redux store 1`] = `null`;

exports[`React-Redux Component PostOverview should render with given state from Redux store 1`] = `
<div
  className="makeStyles-paperRoot-1"
  onClick={[Function]}
>
  <div
    className="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
  >
    <div
      className="makeStyles-postTitle-3"
    >
      Test posts title
    </div>
    <div
      className="makeStyles-userName-4"
      onClick={[Function]}
    >
      Created By 
      Bret
    </div>
  </div>
</div>
`;

exports[`React-Redux Component Posts should render with given state from Redux store 1`] = `null`;
